### PR TITLE
Remove the remaining soft deleteable code

### LINKED ISSUE #4761: What is the status of the SoftDeleteableFilter?
I get some errors on the current master branch due to calls to the softdeleteable filter.

What is the status of the usage of this filter? I would like to clean up the codebase to avoid errors, but I'm not sure what components still are allowed to have a `deletedAt` field.

### PR DESCRIPTION
**Pull request by coudenysj**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | ?
| Deprecations?   | yes
| Related tickets | fixes #4761 and #4461
| License         | MIT

## TODO
- [ ] Cleanup Behat tests
- [x] Check if [DoctrineSluggablePass](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/DoctrineSluggablePass.php) can be removed

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Controller/OrderController.php [hunk 1]
```diff
@@ -46,10 +46,8 @@ public function indexByCustomerAction(Request $request, $id)
 
         // Fetch and cache deleted orders
         $entityManager = $this->container->get('doctrine.orm.entity_manager');
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Controller/OrderController.php]: `$entityManager` is not needed anymore, as well as comment above.

**Comment by coudenysj** [comment]
Looks like TravisCI is having a lot of difficulties?

**Comment by Zales0123** [comment]
@coudenysj yes, sadly it seems to fail with latest chrome version. All should be fixed after merging https://github.com/Sylius/Sylius/pull/4786.

**Comment by coudenysj** [comment]
As far as I can tell, the soft deleteable magic is not tested in any Behat feature file.

We have to wait on #4786 to see if anything breaks, I guess.

**Comment by michalmarcinkowski** [comment]
@coudenysj #4786 was merged, please rebase.

**Comment by coudenysj** [comment]
@michalmarcinkowski Done!

**Comment by michalmarcinkowski** [comment]
Great work! Thanks Jachim! :+1:

**Comment by pjedrzejewski** [comment]
Awesome work Jachim, thank you so much!

**Comment by rvanlaak** [comment]
Great! Thanks, the `softdeleteable` also resulted into errors when the `promotion` bundle was used stand-alone.

**Comment by pjedrzejewski** [comment]
@rvanlaak I am not sure what you mean? Soft-deleteable behavior is gone from Sylius.

**Comment by rvanlaak** [comment]
@pjedrzejewski yes I know, but I'm still on `0.17` and the change was merged after that release. So basically my question is whether, until another version is released, I can just enable it before and disable the filter afterwards?

Oh wait, I see `0.18` is released yesterday. Think I'm going to upgrade right away. Major breaks on `promotion`?

**Comment by rvanlaak** [comment]
The upgrade to `0.18` wasn't easy, got a missing factory error while compiling the container. Probably has something to do with a missing dependency if `promotion` is used stand-alone.

**Comment by michalmarcinkowski** [comment]
@rvanlaak did you resolve this issue?

**Comment by rvanlaak** [comment]
@michalmarcinkowski no we postponed the upgrade for now, but will try again soon. Did hope that the documentation on the upgrade can be improved, given `0.18` has that many changes related to the `promotion` component: https://github.com/Sylius/Sylius/blob/master/CHANGELOG.md

**Comment by rvanlaak** [comment]
We updated from `0.17` to `0.19` today, basically rewriting the `softDelete` functionality took us quite some time. We are fine with removing it from the `Promotion` entity, but we ourselves actually have a use-case for it on the `Coupon` entity. When customers bought a coupon from us but don't want to use it anymore, we should be able to "delete" it.

@michalmarcinkowski are there any plans to simulate the `softDelete` behaviour for the `Coupon` entity, but without using the `DoctrineExtensions`? The only thing about the decision proces I can find on it is you saying that is should be removed in https://github.com/Sylius/Sylius/issues/4761#issuecomment-208777651, but I'm not exactly sure why :)

**Comment by coudenysj** [comment]
@rvanlaak I think a solution could be to use the ToggleableInterface.

**Comment by michalmarcinkowski** [comment]
@rvanlaak if coupon is not used you should be able to delete it, if it was already used it should be archived. See https://github.com/Sylius/Sylius/issues/4140 and our idea for archiving.

**Comment by michalmarcinkowski** [comment]
Btw. for coupons you can simply set "expiresAt" to make it archived :)

**Comment by rvanlaak** [comment]
Ah thanks for referencing #4140, I agree with the problem statement given there 👍 For now I was able to reuse the `deletedAt` field that already exists by overriding the `Coupon` class used by the `promotion` component:

    /**
     * @ORM\Table(name="sylius_promotion_coupon")
     * @ORM\Entity
     *
     * @Gedmo\SoftDeleteable(fieldName="deletedAt")
     */
    class Coupon extends BaseCoupon
    {
        use SoftDeleteableEntity;
    }

After the first alpha is released we will migrate it to `expiresAt`. Wouldn't it be better to "improve" the [`ToggleableTrait`](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Resource/Model/ToggleableTrait.php) and use an `expiresAt` datetime instead of the boolean that is used now? The bool doesn't tell us anything about the coupon's history (read: archive date).

**Comment by michalmarcinkowski** [comment]
Not sure what do you mean? `PromotionCoupon` is not toggleable.

**Comment by rvanlaak** [comment]
Yes that's correct, so I'd like to propose to make [`PromotionCoupon`](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Promotion/Model/PromotionCoupon.php) toggleable and would like to know if you agree on that too. 

The second thing I'd like to mention (aside of making it toggleable itself), is that it would be great if `ToggleableTrait::$enabled` could become a `DateTime` to provide more info about the lifecycle of the entity. In other words: so we can see when the coupon was archived. That would also make archiving in the future possible:

```php
/**
 * @return Boolean
 */
public function isEnabled()
{
	if (null !== $this->enabled) {
		return $this->enabled <= (new \DateTime());
	}

	return false;
}
```

**Comment by michalmarcinkowski** [comment]
That seems redundant. Why not simply use `expiresAt` for this? The logic is already implemented and working.

**Comment by rvanlaak** [comment]
Okay thanks 👍 For our use-case the archiving has a different use-case than the `expiresAt` field would have so I'll stick with my custom implementation for now.

**Pull request merged**